price-estimator/src/solver_rounding_buffer.rs

```diff
@@ -0,0 +1,162 @@
+use core::models::{AccountState, Order, TokenId, TokenInfo};
+use ethcontract::{Address, U256};
+use std::collections::{BTreeMap, HashMap};
+
+// This code is closely related to dex-solver/src/opt/process/Rounding.py .
+// Discussion of motivation happened in https://github.com/gnosis/dex-services/issues/970 .
+
+type Tokens = BTreeMap<TokenId, TokenInfo>;
+
+const MAX_ROUNDING_VOLUME: f64 = 100_000_000_000.0;
+const PRICE_ESTIMATION_ERROR: f64 = 10.0;
+
+fn max_rounding_amount(token_price: f64, fee_token_price: f64) -> f64 {
+    let estimated_price_in_fee_token = token_price / fee_token_price;
+    let max_rounding_amount = MAX_ROUNDING_VOLUME / estimated_price_in_fee_token;
+    max_rounding_amount.max(1.0)
+}
+
+/// Perform the same rounding buffer calculation as our solvers in order to increase the correctness
+/// of our estimates.
+/// All token prices must be nonzero.
+#[allow(dead_code)]
+pub fn apply_rounding_buffer(
+    tokens: &Tokens,
+    orders: &mut Vec<Order>,
+    account_state: &mut AccountState,
+) {
+    let fee_token_price = tokens
+        .get(&TokenId(0))
+        .expect("fee token does not have price")
+        .external_price as f64;
+    assert!(fee_token_price > 0.0, "fee token price is 0");
+
+    // The maximum rounding buffer over all orders from this address selling this token.
+    let mut account_balance_buffers = HashMap::<(Address, TokenId), u128>::new();
+    // Apply rounding buffer to account balances and order sell amounts.
+    for o in orders.iter_mut() {
+        // Compute rounding buffer.
+        let (t_s, t_b) = (TokenId(o.sell_token), TokenId(o.buy_token));
```

Comment:
ubernit: can we please spell out the variable names? I know the python codebase uses abbreviations to make it more look like a mathematical equation/proof but I think it's less cognitive effort to understand this code if I don't have to interpret a one character variable name.

Reply:
The only reason for these names is to keep them the same as in the python code but I don't mind either way. Definitely wouldn't have used them if this wasn't moved over from there. I'll change it.